chain-impl-mockchain/src/ledger/ledger.rs

```diff
@@ -1140,6 +1170,51 @@ impl Ledger {
         Ok(self)
     }
 
+    pub fn apply_encrypted_vote_tally<'a>(
+        mut self,
+        tally: &certificate::EncryptedVoteTally,
+        bad: &TransactionBindingAuthData<'a>,
+        sig: certificate::EncryptedVoteTallyProof,
+    ) -> Result<Self, Error> {
+        if sig.verify(tally.tally_type(), bad) == Verification::Failed {
+            return Err(Error::VoteTallyProofFailed);
+        }
+
+        let stake = StakeControl::new_with(&self.accounts, &self.utxos);
+
+        let mut f = |_action: &VoteAction| ();
+
+        self.votes = self.votes.apply_encrypted_vote_tally(
+            self.date(),
+            &stake,
+            &self.governance,
+            tally,
+            sig,
+            &mut f,
+        )?;
+
+        Ok(self)
+    }
+
+    pub fn apply_private_vote_tally<'a>(
+        self,
+        tally: &certificate::PrivateVoteTally,
+        bad: &TransactionBindingAuthData<'a>,
+        sig: certificate::TallyProof,
+    ) -> Result<Self, Error> {
+        if sig.verify(tally.tally_type(), bad) == Verification::Failed {
+            return Err(Error::VoteTallyProofFailed);
+        }
+
+        let _stake = StakeControl::new_with(&self.accounts, &self.utxos);
+
+        let mut _f = |_action: &VoteAction| ();
+
+        // TODO: Decrypt shares here
```

Comment:
I'm missing how and where should we decrypt the shares.

Reply:
We should not really decrypt them. Just feed them and the tally state (IIRC it is the same for all users) to the `result` function. https://github.com/input-output-hk/chain-libs/blob/master/chain-vote/src/lib.rs#L129